pyatv/support/dns.py

```diff
@@ -0,0 +1,344 @@
+"""Processing functions for raw DNS messages."""
+import enum
+import io
+import logging
+import struct
+import typing
+from ipaddress import IPv4Address
+
+from pyatv.support.collections import CaseInsensitiveDict
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def unpack_stream(fmt: str, buffer: typing.BinaryIO) -> typing.Tuple:
+    """Unpack a tuple from a binary stream, much like `struct.unpack`."""
+    size = struct.calcsize(fmt)
+    return struct.unpack(fmt, buffer.read(size))
+
+
+def qname_encode(name: str) -> bytes:
+    """Encode QNAME without using name compression."""
+    encoded = bytearray()
+    labels = name.split(".")
+    # Ensure there's an empty label for the root domain
+    if labels[-1] != "":
+        labels.append("")
+    for label in labels:
+        encoded_label = label.encode("idna")
+        encoded_length = len(encoded_label)
+        # Length of the encoded label, in bytes, but a maximum of 63
+        # The maximum is 63 as the upper two bits are used as a flag for name
+        # compression.
+        encoded.append(min(encoded_length, 63))
+        if encoded_length == 0:
+            # If we've reached an empty label, assume this is the last component.
+            # Empty labels (two periods right after each other) isn't legal anyways.
+            break
+        # TODO: Should an error be raised if a label is too long?
```

Comment:
I guess we should not end up here, but maybe log a warning or at least debug about it?

Reply:
Yeah, I wasn't sure if raising a `ValueError` would be more appropriate than just logging it. If this was part of a more DNS focused-library maybe, but this is just a supporting module for Apple TV support. Speaking of which, I think I remember seeing macOS truncate names when they got too long too (but I haven't had such a long hostname in a while).